### PR TITLE
Fix ValueError when passing tf.data.Dataset to model.fit

### LINKED ISSUE #24520: "ValueError: Cannot take the length of Shape with unknown rank". error when passing tf.data.Dataset tensors to model.fit
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): "18.04.1 LTS (Bionic Beaver)"
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: no
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): tf.VERSION = 1.12.0
- Python version: python3.6
- Bazel version (if compiling from source): no
- GCC/Compiler version (if compiling from source): no
- CUDA/cuDNN version: cuda9.0 with cuDNN 7.4.1
- GPU model and memory: GTX 1080 with 8 GB

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

**Describe the current behavior**
I am trying to pass the tfrecords read through tf.data.Dataset api into the model.fit .
Since the images could be of different sizes, I am storing the image shapes into tfrecords itself which are 
later on read and applied to the img data using  `tf.reshape` . But the tensorflow.keras is unable to determine the shape of this image data at this stage and throws the error.

```
def _parse_function(proto):
    keys_to_features = {"im_path": tf.FixedLenSequenceFeature([], tf.string, allow_missing=True),
                        "im_shape": tf.FixedLenSequenceFeature([], tf.int64, allow_missing=True),
                        "im_arr": tf.FixedLenSequenceFeature([], tf.string, allow_missing=True),
                        "label": tf.FixedLenSequenceFeature([], tf.int64, allow_missing=True),
                        }

    parsed_features = tf.parse_single_example(serialized=proto, features=keys_to_features)
    parsed_features['im_arr'] = parsed_features['im_arr'][0]
    parsed_features['label'] = parsed_features['label'][0]
    parsed_features['im_arr'] = tf.decode_raw(parsed_features['im_arr'], tf.uint8)
    parsed_features['im_arr'] = tf.reshape(parsed_features['im_arr'], parsed_features['im_shape'])

    return parsed_features['im_arr'], parsed_features['label']
```

 The error thrown is as follows : 

	Traceback (most recent call last):
	  File "issue/IssueScript.py", line 75, in <module>
	    verbose=1)
	  File "opt/github/example-classification/env/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 1536, in fit
	    validation_split=validation_split)
	  File "opt/github/example-classification/env/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 992, in _standardize_user_data
	    class_weight, batch_size)
	  File "opt/github/example-classification/env/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 1117, in _standardize_weights
	    exception_prefix='input')
	  File "opt/github/example-classification/env/lib/python3.6/site-packages/tensorflow/python/keras/engine/training_utils.py", line 284, in standardize_input_data
	    data = [standardize_single_array(x) for x in data]
	  File "opt/github/example-classification/env/lib/python3.6/site-packages/tensorflow/python/keras/engine/training_utils.py", line 284, in <listcomp>
	    data = [standardize_single_array(x) for x in data]
	  File "opt/github/example-classification/env/lib/python3.6/site-packages/tensorflow/python/keras/engine/training_utils.py", line 218, in standardize_single_array
	    if x.shape is not None and len(x.shape) == 1:
	  File "opt/github/example-classification/env/lib/python3.6/site-packages/tensorflow/python/framework/tensor_shape.py", line 579, in __len__
	    raise ValueError("Cannot take the length of Shape with unknown rank.")
	ValueError: Cannot take the length of Shape with unknown rank.

So as a debugging step, I removed the length check present in the `standardize_single_array` function by changing the check as (note the `False and` part which bypasses the length check)

```def standardize_single_array(x):
  if x is None:
    return None
  if False and (x.shape is not None and len(x.shape) == 1):
    if tensor_util.is_tensor(x):
      return array_ops.expand_dims(x, axis=1)
    else:
      return np.expand_dims(x, 1)
  return x
```
Then I get the following error 

	Traceback (most recent call last):
	  File "issue/IssueScript.py", line 75, in <module>
	    verbose=1)
	  File "opt/github/example-classification/env/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 1536, in fit
	    validation_split=validation_split)
	  File "opt/github/example-classification/env/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 992, in _standardize_user_data
	    class_weight, batch_size)
	  File "opt/github/example-classification/env/lib/python3.6/site-packages/tensorflow/python/keras/engine/training.py", line 1154, in _standardize_weights
	    exception_prefix='target')
	  File "opt/github/example-classification/env/lib/python3.6/site-packages/tensorflow/python/keras/engine/training_utils.py", line 323, in standardize_input_data
	    'with shape ' + str(data_shape))
	ValueError: Error when checking target: expected activation_4 to have 2 dimensions, but got array with shape (None,)

I did the same with the above error. I removed the check present at line 323 by commenting out the length check as follows.
```
        """
        if len(data_shape) != len(shape):
          raise ValueError('Error when checking ' + exception_prefix +
                           ': expected ' + names[i] + ' to have ' +
                           str(len(shape)) + ' dimensions, but got array '
                           'with shape ' + str(data_shape))
        """
```
Now the training proceeds smoothly without error. I believe there is issue with tf.reshape when tensors are supplied as a shape to the function.

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
Code : https://github.com/dineshdharme/tensorflow-issue1
Just run : `python3 issue/IssueScript.py`

I have also added a tfrecords generating script `tfrecords_utils.py` which you can call by
To generate tfrecords file using the image data present in the data folder : 
`python3 issue/tfrecords_utils.py`

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #24520 where
passing tf.data.Dataset to model.fit might result in:
```
ValueError: Cannot take the length of Shape with unknown rank.
```

This fix fixes the error by replacing len(x.shape) with x.shape.ndims

This fix fixes #24520.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/python/keras/engine/training_utils.py [hunk 1]
```diff
@@ -228,9 +228,12 @@ def standardize_single_array(x, expected_shape=None):
   """Expand data of shape (x,) to (x, 1), unless len(expected_shape)==1."""
   if x is None:
     return None
+  if tensor_util.is_tensor(x):
+    x_shape_ndims = x.shape.ndims if x.shape is not None else None
```
Comment (fchollet) [inline tensorflow/python/keras/engine/training_utils.py]: Prefer using `shape.rank`

**Review by fchollet — DISMISSED** [review]
Thanks for the PR. LGTM (one minor nit)

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @fchollet: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by ymodak** [comment]
@yongtang Can you resolve the merge conflicts? Thanks!

File: tensorflow/python/keras/engine/training_utils.py [hunk 2]
```diff
@@ -228,9 +228,12 @@ def standardize_single_array(x, expected_shape=None):
   """Expand data of shape (x,) to (x, 1), unless len(expected_shape)==1."""
   if x is None:
     return None
+  if tensor_util.is_tensor(x):
+    x_shape_ndims = array_ops.rank(x)
```
Comment (pavithrasv) [inline tensorflow/python/keras/engine/training_utils.py]: Can we add a unit test covering this use case (the issue reported in  #24520)?

**Comment by pateli18** [comment]
@yongtang any chance you'll be able to fix this soon (it would be really helpful to me on my current project)? Thanks so much for doing this!

**Comment by rthadur** [comment]
@yongtang gentle ping

**Comment by gbaned** [comment]
@yongtang Any update on above comment please?

**Comment by dd1923** [comment]
@yongtang Could you please check on line 756 , too? It raises similar error when one is training with class_weights and y is a tensor from tf.dataset. The error reads something like this:

File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/keras/engine/training_utils.py", line 756, in standardize_weights
    if len(y.shape) > 2:
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/framework/tensor_shape.py", line 745, in __len__
    raise ValueError("Cannot take the length of shape with unknown rank.")

**Comment by dd1923** [comment]
@yongtang  any updates? As I said above, this is affecting in other places too. I am not able to train a model with class_weights and tf.dataset

**Comment by sallamander** [comment]
@dd1923 for what it's worth, I ran into the same error when using `tf.Dataset.from_generator`, and was able to fix the issue by passing in an argument for [`output_shapes`](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#args_7). I didn't even have to specify the exact dimensions - using something like `tf.TensorShape([None, None, None])` worked in the case of using 3 dimensional images. While the `from_tensor_slices` method doesn't have have the same parameter, I wonder if there is an analogous way to set the shape (even if filled with `None`, but having the right dimensionality) that might resolve your issue until this PR is merged.

So, I have an imagenet generator outputting (image, label) of shape ((224, 224, 3), (1, )).

Original code that leads to that same error being thrown:

```
dataset = tf.Dataset.from_generator(
    generator, output_types=('float32', 'uint8')
)
```

Fixed code that prevents that error:

```
dataset = tf.Dataset.from_generator(
    generator, output_types=('float32', 'uint8'),
    output_shapes=(tf.TensorShape((None, None, None)), tf.TensorShape((1, )))
)

**Comment by dd1923** [comment]
@sallamander that;s a good discovery. I gave up on tf.dataset lately. I switched to keras sequences and now everything just works. I think Tensorflow hastily declared support for tf.dataset in Keras APIs without first making sure that the underlying code is ready for that. It was a nightmare to get a few basic things work.

**Comment by gbaned** [comment]
@yongtang can you please resolve conflicts? Thanks!

**Comment by rthadur** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Comment by jordanra-miso** [comment]
Is it possible to get this actually reopened and merged? The conflicts are relatively trivial of a fix, and this is still a bug in Tensorflow 1.14. (Unknown whether this error occurs for Tensorflow 2.0, our team has had no luck with getting 2.0 to work in our codebase).

**Comment by rthadur** [comment]
@jordanra-miso please open a new PR with this changes , We will help merge the PR. Thank you

**Comment by jordanra-miso** [comment]
@rthadur My PR (#33895) has been made, with updated `shape.rank` usage to conform to @fchollet's comments.

**Pull request closed without merging**